### PR TITLE
KOGITO-7172 [SWF Editor] Analzyzing and Defining a API to manage cursor in VSCode

### PR DESCRIPTION
**Pull request by fantonangeli**
Jira: https://issues.redhat.com/browse/KOGITO-7172

Note 1:
There is a minor bug that appears very randomly in VSCode only:
When the user click on a node and then double/triple click on another node, VSCode opens a new text editor on the right side.

Note 2:
To locate the line of the selected node on the text editor, I just use regular expressions. 
Because in `serverless-workflow-language-service` we already have `jsonc-parser`, I made some attempts to use it to ensure that the line belongs to a node inside the `states` array, but didn't manage to get the line number.
Here a codepen with my attempt to use jsonc-parser: https://codesandbox.io/s/zen-williamson-keyem4?file=/src/index.js

kogito-7172

### CODE REVIEW

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (caponetto) [inline packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts:1]: Let's avoid having a `utils` folder. 
Please create a folder with a meaningful name and place this code in an `index.ts` file.

File: packages/serverless-workflow-language-service/tests/findPositionByStateName.test.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    type: inject
+    data:
+      result: Hello World!
+    end: true
+`;
+
+describe("findPositionByStateName", () => {
+  it("should return null with wrong inputs", () => {
+    // @ts-ignore
+    expect(findPositionByStateName(null, null)).toBe(null);
```
Comment (caponetto) [inline packages/serverless-workflow-language-service/tests/findPositionByStateName.test.ts:68]: You can use `toBeNull` where applicable.

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts (hunk 1 above)
Comment (fantonangeli) [reply packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts:1]: I put it in a separated .ts file because I expect we will need to extend it in future, eg. writing some checks on the type of node...
Anyway I agree to put it in a folder with a better name. WDYT about `/src/editor/index.ts`? Would you have better names for it?

File: packages/serverless-workflow-text-editor/src/editor/textEditor/SwfTextEditor.tsx [hunk 3]
```diff
@@ -109,8 +110,17 @@ const RefForwardingSwfTextEditor: React.ForwardRefRenderFunction<SwfTextEditorAp
 
   useSubscription(
     editorEnvelopeCtx.channelApi?.notifications.kogitoSwfLanguageService__moveCursorToNode,
-    (nodeName: string) => {
-      console.log("nodeName", nodeName);
+    ({ nodeName }: { nodeName: string }) => {
+      const targetPosition = findPositionByStateName(content, nodeName, "JSON");
```
Comment (caponetto) [inline packages/serverless-workflow-text-editor/src/editor/textEditor/SwfTextEditor.tsx:114]: Please resolve the 3rd arg from the file extension. This way, the code is ready for yaml files.

File: packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts [hunk 4]
```diff
@@ -118,7 +119,38 @@ export async function setupDiagramEditorControls(args: {
 
   args.context.subscriptions.push(
     vscode.commands.registerCommand(COMMAND_IDS.moveCursorToNode, async ({ nodeName }: { nodeName: string }) => {
-      console.log("vscode moveCursorToNode command called with nodeName", nodeName);
+      const textEditor = vscode.window.visibleTextEditors[0] || null;
+
+      if (!textEditor || vscode.window.visibleTextEditors.length > 1) {
```
Comment (caponetto) [inline packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts:124]: The feature does not work if I have more than one text editor opened?

File: packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts [hunk 5]
```diff
@@ -118,7 +119,38 @@ export async function setupDiagramEditorControls(args: {
...
+        console.debug("TextEditor not found");
+        return;
+      }
+
+      const resourceUri = textEditor.document.uri;
+      const targetPosition = findPositionByStateName(textEditor.document.getText(), nodeName);
```
Comment (caponetto) [inline packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts:130]: Please resolve the extension and pass it as the 3rd arg so the code is ready for yaml files.

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  if (!fullText || !stateName) {
+    return null;
+  }
+
+  const fullTextSplit = fullText.split("\n");
+  let nameRegExp = new RegExp(`"name"\\s*:\\s*"${stateName}"`);
+
+  if (fileType === "YAML") {
+    nameRegExp = new RegExp(`name\\s*:\\s*${stateName}`);
+  }
```
Comment (caponetto) [inline packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts:44]: I would set `nameRegExp` as const and use a ternary assignment.

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+  for (let lineNum = 0, end = fullTextSplit.length; lineNum < end; lineNum++) {
```
Comment (caponetto) [inline packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts:46]:
```suggestion
  for (let lineNum = 0; lineNum < fullTextSplit.length; lineNum++) {
```

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts (hunk 1 above)
Comment (caponetto) [reply packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts:1]: I think for now an `editor` folder looks good!

File: packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts (hunk 4 above)
Comment (fantonangeli) [reply packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts:124]: It works if you have more files opened, but the JSON/YAML must be visible.
I just limited it not to work in case you have the window splitter in more than two groups like this case
Screenshot from 2022-06-07 18-16-54
:

File: packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts [hunk 8]
```diff
@@ -118,7 +119,38 @@ export async function setupDiagramEditorControls(args: {
...
+
+      if (targetPosition === null) {
+        return;
+      }
+
+      await vscode.commands.executeCommand("vscode.open", resourceUri, {
```
[the remark this answers is not in the document]
Comment (fantonangeli) [reply packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts:136]: Yes, I didn't find a more specific one. Would you have any suggestions?

File: packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts (hunk 7 above)
Comment (caponetto) [reply packages/serverless-workflow-language-service/src/utils/findPositionByStateName.ts]: I think there is some debate on this matter and seeing a performance improvement would depend on the context.
But I'm fine with this approach as well.

**Review by caponetto — COMMENTED** [review]
Thanks, @fantonangeli!
Can you spend some more time on the two notes you raised?
I'll take a look at them asap.

File: packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts (hunk 4 above)
Comment (fantonangeli) [reply packages/vscode-extension-serverless-workflow-editor/src/extension/setupDiagramEditorControls.ts]: Now I implemented the ability to work with multiple tabs open.
https://github.com/caponetto/kie-tools/pull/29/commits/62431f2046d7c31788a9a1dd934400a5ff9a7eb0
kogito-7172-3split

**Review by caponetto — APPROVED** [review]
Excellent work, @fantonangeli! I agree.
@yesamer mind taking a final look?

**Review by yesamer — APPROVED** [review]
@fantonangeli @caponetto Excellent work, the functionality. the code looks good to me, so I agree to merge it. I just report a couple of corner cases you should evaluate and possibly manage in future PRs. Thank you!

File: packages/serverless-workflow-language-service/tests/findPositionByStateName.test.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+ * limitations under the License.
+ */
+
+import { findPositionByStateName, FileLanguage } from "@kie-tools/serverless-workflow-language-service/dist/editor";
+
+const mockJson = `{
+  "id":"helloworld",
+  "version":"1.0",
+  "specVersion":"0.8",
+  "name":"Hello World Workflow",
```
[the remark this answers is not in the document]
Comment (fantonangeli) [reply packages/serverless-workflow-language-service/tests/findPositionByStateName.test.ts:23]: Totally agree, those are the checks that I wanted to implement in the "Note 2".
I opened a Jira for this: https://issues.redhat.com/browse/KOGITO-7367

File: packages/serverless-workflow-language-service/src/editor/index.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @param fileName the filename or path
+ * @returns the file language, null if found
+ */
+export const getFileLanguage = (fileName: string): FileLanguage | null => {
+  if (!fileName) {
+    return null;
+  }
+
+  if (fileName.endsWith(".sw.json")) {
```
[the remark this answers is not in the document]
Comment (caponetto) [reply packages/serverless-workflow-language-service/src/editor/index.ts:78]: Good catch! This code was moved from somewhere else but it is worth treating this case.
I guess `fileName.toLowerCase().endsWith...` would be enough.
